src/lib.rs

```diff
@@ -198,13 +221,59 @@ fn num_bits<T>() -> usize {
     size_of::<T>() * 8
 }
 
-fn i_log2<T: PrimInt + Unsigned>(n: T) -> usize {
-    debug_assert!(n > T::zero());
-    let mut bits = num_bits::<T>() - 1;
-    while n & (T::one() << bits) == T::zero() {
-        bits -= 1;
+/// Convert (possibly signed) `x` of type `T` into an unsigned `StorageT`.
+fn abs<T, StorageT>(x: T) -> StorageT
+where
+    T: 'static + AsPrimitive<StorageT> + FromPrimitive + Ord + PrimInt + ToPrimitive,
+    StorageT: AsPrimitive<T> + PrimInt + Unsigned,
+{
+    debug_assert!(size_of::<StorageT>() >= size_of::<T>());
+    if x >= T::zero() {
+        x.as_()
+    } else if x == T::min_value() {
+        T::max_value().as_() + StorageT::one()
+    } else {
+        (T::zero() - x).as_()
+    }
+}
+
+/// Return the delta between `min` and `max` as an unsigned integer (i.e. delta(-2, 2) == 4).
+fn delta<T, StorageT>(min: T, max: T) -> StorageT
+where
+    T: 'static + AsPrimitive<StorageT> + FromPrimitive + Ord + PrimInt + ToPrimitive,
+    StorageT: AsPrimitive<T> + PrimInt + Unsigned,
+{
+    debug_assert!(size_of::<StorageT>() >= size_of::<T>());
+    if min >= T::zero() {
+        (max - min).as_()
+    } else if min < T::zero() && max < T::zero() {
+        abs(max) - abs(min)
+    } else {
+        debug_assert!(min < T::zero());
+        max.as_() + abs(min)
     }
-    bits + 1
+}
+
+/// Given a (possibly signed) minimum value `min` and an absolute delta `d`, return a (possibly)
+/// signed inverted value (i.e. inv_delta(-2, 4) == 2).
+fn inv_delta<T, StorageT>(min: T, d: StorageT) -> T
+where
+    T: 'static + AsPrimitive<StorageT> + FromPrimitive + Ord + PrimInt + ToPrimitive,
+    StorageT: AsPrimitive<T> + PrimInt + Unsigned,
+{
+    debug_assert!(size_of::<StorageT>() >= size_of::<T>());
+    if d <= T::max_value().as_() {
```

Comment:
This is another tricky one :/ `d` is `StorageT`. So if `T` is a signed type this might not be true. As an example, if `T` is i8 and `StorageT` u8, `T::max_value()` == 127, but `StorageT::max_value()` == 255, so you can get values of `d` that are bigger than `T::max_value()`.